compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -598,39 +621,85 @@ class ClassfileParser(
       // the classpath would *not* end up here. A class not found is signaled
       // with a `FatalError` exception, handled above. Here you'd end up after a NPE (for example),
       // and that should never be swallowed silently.
-      report.warning("Caught: " + ex + " while parsing annotations in " + in.file)
+      report.warning("Caught: " + ex + " while parsing annotations in " + classfile)
       if (ctx.debug) ex.printStackTrace()
 
       None // ignore malformed annotations
   }
 
-  def parseAttributes(sym: Symbol, symtype: Type, isVarargs: Boolean = false)(using Context): Type = {
-    var newType = symtype
+  /** A completer for attributes
+   *
+   *  Top-level classes complete attributes eagerly, while members complete lazily.
+   *
+   *  @note We cannot simply store the bytes of attributes, as the bytes may
+   *  contain references to the constant pool, where the constants are loaded
+   *  lazily.
+   */
+  class AttributeCompleter(sym: Symbol) {
```

Comment:
Moving this class outside of ClassfileParser would also be useful for the same reason, but in fact it won't work currently as we're calling methods from the enclosing class (`sigToType`, `convertTo`) and so we're keeping a reference to it. Can we move these methods to the companion object to avoid this problem?

Reply:
Yes, as you noticed, it depends on instance methods like `sigToType`, the same as `MemberCompleter`, therefore indirectly depend on `innerClasses` and etc. I think there is a refactoring opportunity to clean up things. As it will make classfile parser very different from it is now, I'd reserve it for another PR to make this PR simpler.